src/main/java/spoon/reflect/path/impl/CtRolePathElement.java

```diff
@@ -106,14 +54,65 @@ public CtRole getRole() {
 
 	@Override
 	public String toString() {
-		return STRING + role.toString() + getParamString();
+		return STRING + getRole().toString() + getParamString();
+	}
+
+	public CtElement getFromSet(Set set, String name) throws CtPathException {
+		for (Object o: set) {
+			if (o instanceof CtNamedElement) {
+				if (((CtNamedElement) o).getSimpleName().equals(name)) {
+					return (CtElement) o;
+				}
+			} else if (o instanceof CtReference) {
+				if (((CtReference) o).getSimpleName().equals(name)) {
+					return (CtElement) o;
+				}
+			} else {
+				throw new CtPathException();
+			}
+		}
+		throw new CtPathException();
 	}
 
 	@Override
 	public Collection<CtElement> getElements(Collection<CtElement> roots) {
-		RoleVisitor visitor = new RoleVisitor();
-		visitor.scan(roots);
-		return visitor.matchs;
+		Collection<CtElement> matchs = new LinkedList<>();
+		for (CtElement root : roots) {
+			RoleHandler roleHandler = RoleHandlerHelper.getOptionalRoleHandler(root.getClass(), getRole());
+			if (roleHandler != null) {
+				switch (roleHandler.getContainerKind()) {
+					case SINGLE:
+						matchs.add(roleHandler.getValue(root));
+						break;
+
+					case LIST:
+						if (getArguments().containsKey("index")) {
+							int index = Integer.parseInt(getArguments().get("index"));
+							matchs.add((CtElement) roleHandler.asList(root).get(index));
+						}
+						break;
+
+					case SET:
+						if (getArguments().containsKey("name")) {
+							String name = getArguments().get("name");
+							try {
+								matchs.add(getFromSet(roleHandler.asSet(root), name));
+							} catch (CtPathException e) {
+								//System.err.println("[ERROR] Element not found for name: " + name);
+								//No element found for name.
+							}
+						}
+						break;
+
+					case MAP:
+						if (getArguments().containsKey("key")) {
+							String name = getArguments().get("key");
+							matchs.add((CtElement) roleHandler.asMap(root).get(name));
```

Comment:
I guess it is not correct to add null into matchs, when map doesn't contain a key

Reply:
Indeed. I'll fix this too.